File: python/examples/fraud_prevention.py
"""
Fraud Prevention Example (Python)

Shows how a fintech backend evaluates transaction risk using
TrustLayer's Fraud Shield.

Run:
    TRUSTLAYER_SECRET_KEY=tl_secret_xxx python examples/fraud_prevention.py
"""

import os
from trustlayer import TrustLayer

API_KEY = os.environ.get("TRUSTLAYER_SECRET_KEY", "tl_secret_example")
API_URL = os.environ.get("TRUSTLAYER_API_URL", "http://localhost:8080")


def evaluate_transaction(
    user_id: str,
    amount: float,
    currency: str = "USD",
    is_new_device: bool = False,
    location_changed: bool = False,
) -> None:
    print(f"\nEvaluating transaction for user {user_id}")
    print(f"  Amount: {currency} {amount:,.2f}")

    with TrustLayer(api_key=API_KEY, api_url=API_URL) as client:
        # 1. Create a transaction session
        session = client.create_session(
            type="transaction",
            user_id=user_id,
            modules=["fraud", "anomaly"],
            metadata={"amount": amount, "currency": currency},
        )

        # 2. Send contextual signals
        if is_new_device:
            session.track_event("device_change", {"is_new": True})

        if location_changed:
            session.track_event(
                "suspicious_activity",
                {"reason": "location_change", "vpn_detected": True},
            )

        if amount > 5000:
            session.track_event(
                "custom",
                {
                    "signal_type": "transaction",
                    "amount": amount,
                    "currency": currency,
                    "high_value": True,
                },
            )

        # 3. Use FraudShield for evaluation
        fraud = client.fraud(session)
        result = fraud.evaluate(
            transaction_amount=amount,
            transaction_currency=currency,
            user_id=user_id,
        )

        # 4. Print decision
        print("\n─── Fraud Assessment ─────────────────────────────")
        print(f"  Fraud Probability : {result.risk_score.risk_score / 100:.1%}")
        print(f"  Risk Score        : {result.risk_score.risk_score:.0f}/100")
        print(f"  Recommendation    : {result.recommendation.upper()}")
        if result.risk_score.primary_factors:
            print("  Risk Factors:")
            for f in result.risk_score.primary_factors:
                print(f"    • {f}")
        print("──────────────────────────────────────────────────")

        # 5. Complete
        session.complete()


if __name__ == "__main__":
    # Low-risk transaction
    evaluate_transaction(
        user_id="user_established_001",
        amount=49.99,
        is_new_device=False,
        location_changed=False,
    )

    # High-risk transaction
    evaluate_transaction(
        user_id="user_new_002",
        amount=8500.00,
        is_new_device=True,
        location_changed=True,
    )

File: examples/fraud-prevention.ts
/**
 * Fraud Prevention Example
 *
 * Shows how a fintech/e-commerce platform evaluates transaction risk
 * using TrustLayer's Fraud Shield from a Node.js backend.
 *
 * Usage:
 *   npx ts-node examples/fraud-prevention.ts
 */

import { TrustLayer } from "../src";

interface Transaction {
  id: string;
  userId: string;
  amount: number;
  currency: string;
  deviceId?: string;
  ipAddress?: string;
  isNewDevice?: boolean;
  locationChanged?: boolean;
}

async function evaluateTransaction(tx: Transaction): Promise<void> {
  const tl = TrustLayer.initialize({
    apiKey: process.env["TRUSTLAYER_SECRET_KEY"] ?? "tl_secret_example",
    apiUrl: process.env["TRUSTLAYER_API_URL"] ?? "http://localhost:8080",
    debug: false,
  });

  console.log(`\nEvaluating transaction ${tx.id}...`);
  console.log(`  User:    ${tx.userId}`);
  console.log(`  Amount:  ${tx.currency} ${tx.amount.toLocaleString()}`);

  // ── 1. Create a transaction session ───────────────────────────────────────
  const session = await tl.createSession({
    type: "transaction",
    userId: tx.userId,
    modules: ["fraud", "anomaly"],
    metadata: { transaction_id: tx.id },
  });

  // ── 2. Send contextual signals ────────────────────────────────────────────
  if (tx.isNewDevice) {
    await session.trackEvent("device_change", {
      device_id: tx.deviceId,
      is_new: true,
    });
  }

  if (tx.locationChanged) {
    await session.trackEvent("suspicious_activity", {
      reason: "location_change",
      vpn_detected: true,
    });
  }

  if (tx.amount > 5000) {
    await session.trackEvent("custom", {
      signal_type: "transaction",
      amount: tx.amount,
      currency: tx.currency,
      high_value: true,
    });
  }

  // ── 3. Evaluate fraud risk ─────────────────────────────────────────────────
  const fraudShield = tl.fraud(session);
  const result = await fraudShield.evaluate({
    transactionAmount: tx.amount,
    transactionCurrency: tx.currency,
    userId: tx.userId,
    transactionId: tx.id,
  });

  // ── 4. Make decision ───────────────────────────────────────────────────────
  console.log("\n─── Fraud Assessment ─────────────────────────────────────");
  console.log(`  Fraud Probability: ${(result.fraudProbability * 100).toFixed(1)}%`);
  console.log(`  Risk Score:        ${result.riskScore}/100`);
  console.log(`  Recommendation:    ${result.recommendation.toUpperCase()}`);

  if (result.factors.length > 0) {
    console.log(`  Risk Factors:`);
    result.factors.forEach((f) => console.log(`    • ${f}`));
  }

  // Application-level decision
  switch (result.recommendation) {
    case "allow":
      console.log("\n  ✓ TRANSACTION APPROVED");
      break;
    case "review":
      console.log("\n  ⚠ TRANSACTION FLAGGED FOR REVIEW — Require step-up auth");
      break;
    case "block":
      console.log("\n  ✗ TRANSACTION BLOCKED");
      break;
  }
  console.log("──────────────────────────────────────────────────────────");

  // ── 5. Complete session ────────────────────────────────────────────────────
  await session.complete();
}

// ── Example transactions ───────────────────────────────────────────────────────

async function main() {
  // Low-risk transaction
  await evaluateTransaction({
    id: "txn_001",
    userId: "user_established_123",
    amount: 49.99,
    currency: "USD",
    isNewDevice: false,
    locationChanged: false,
  });

  // High-risk transaction
  await evaluateTransaction({
    id: "txn_002",
    userId: "user_new_456",
    amount: 8500,
    currency: "USD",
    deviceId: "new-device-xyz",
    isNewDevice: true,
    locationChanged: true,
  });
}

main().catch(console.error);
